tasks/abs.rb

```diff
@@ -9,157 +9,171 @@
+    headers = { 'X-AUTH-TOKEN' => token_from_fogfile('abs'), 'Content-Type' => 'application/json' }
+    payload = { 'job_id' => job_id,
+                'hosts' => [{ 'hostname' => node_name, 'type' => platform, 'engine' => 'vmpooler' }] }
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Post.new(uri.request_uri, headers)
+    request.body = payload.to_json
 
-  targets_to_remove = []
-  inventory_hash['groups'].each do |group|
-    group['targets'].each do |node|
-      targets_to_remove.push(node['uri']) if node['facts']['job_id'] == job_id
+    reply = http.request(request)
+    raise "Error: #{reply}: #{reply.message}" unless reply.code == '200'
+
+    targets_to_remove.each do |target|
+      remove_node(inventory_hash, target)
     end
+    File.open(inventory_full_path, 'w') { |f| f.write inventory_hash.to_yaml }
+    { status: 'ok', removed: targets_to_remove }
   end
-  uri = URI.parse('https://abs-prod.k8s.infracore.puppet.net/api/v2/return')
-  headers = { 'X-AUTH-TOKEN' => token_from_fogfile('abs'), 'Content-Type' => 'application/json' }
-  payload = { 'job_id' => job_id,
-              'hosts' => [{ 'hostname' => node_name, 'type' => platform, 'engine' => 'vmpooler' }] }
-  http = Net::HTTP.new(uri.host, uri.port)
-  http.use_ssl = true
-  request = Net::HTTP::Post.new(uri.request_uri, headers)
-  request.body = payload.to_json
-
-  reply = http.request(request)
-  raise "Error: #{reply}: #{reply.message}" unless reply.code == '200'
-
-  targets_to_remove.each do |target|
-    remove_node(inventory_hash, target)
+
+  def task(action:, platform: nil, node_name: nil, inventory: nil, vars: nil, **_kwargs)
+    inventory_location = sanitise_inventory_location(inventory)
+    result = provision(platform, inventory_location, vars) if action == 'provision'
+    result = tear_down(node_name, inventory_location) if action == 'tear_down'
+    result
   end
-  File.open(inventory_full_path, 'w') { |f| f.write inventory_hash.to_yaml }
-  { status: 'ok', removed: targets_to_remove }
-end
 
-params = JSON.parse(STDIN.read)
-platform = params['platform']
-action = params['action']
-node_name = params['node_name']
-inventory_location = sanitise_inventory_location(params['inventory'])
-vars = params['vars']
-raise 'specify a node_name when tearing down' if action == 'tear_down' && node_name.nil?
-raise 'specify a platform when provisioning' if action == 'provision' && platform.nil?
-unless node_name.nil? ^ platform.nil?
-  case action
-  when 'tear_down'
-    raise 'specify only a node_name, not platform, when tearing down'
-  when 'provision'
-    raise 'specify only a platform, not node_name, when provisioning'
-  else
-    raise 'specify only one of: node_name, platform'
+  def self.run
+    params = JSON.parse(STDIN.read)
+    params.transform_keys! { |k| k.to_sym }
+    action = params[:action]
+    node_name = params[:node_name]
+    platform = params[:platform]
```

Comment:
```suggestion
    action, node_name, platform = params.values_at(:action, :node_name, :platform)
```

Can make this more concise with https://ruby-doc.org//core-2.2.0/Hash.html#method-i-values_at plus array destructuring.